pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -61,60 +62,96 @@ func NewScheduler(kubeClient client.Client) *Scheduler {
 	}
 }
 
-func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) (nodes []*Node, err error) {
+func (s *Scheduler) Solve(ctx context.Context, constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
-	constraints := provisioner.Spec.Constraints.DeepCopy()
 
 	sort.Slice(pods, byCPUAndMemoryDescending(pods))
 	sort.Slice(instanceTypes, byPrice(instanceTypes))
 
 	topology := NewTopology(s.kubeClient, constraints)
-
-	nodeSet, err := NewNodeSet(ctx, topology, constraints, instanceTypes, s.kubeClient)
-	if err != nil {
-		return nil, fmt.Errorf("constructing nodeset, %w", err)
-	}
-
 	if err := topology.Update(ctx, pods...); err != nil {
 		return nil, fmt.Errorf("tracking topology counts, %w", err)
 	}
 
-	schedulingErrors := map[*v1.Pod]error{}
-	unschedulablePods := append([]*v1.Pod{}, pods...)
+	daemonOverhead, err := s.getDaemonOverhead(ctx, constraints)
+	if err != nil {
+		return nil, err
+	}
 
-	previousUnschedulableCount := 0
 	// We loop and retrying to schedule to unschedulable pods as long as we are making progress.  This solves a few
 	// issues including pods with affinity to another pod in the batch. We could topo-sort to solve this, but it wouldn't
 	// solve the problem of scheduling pods where a particular order is needed to prevent a max-skew violation. E.g. if we
 	// had 5xA pods and 5xB pods were they have a zonal topology spread, but A can only go in one zone and B in another.
 	// We need to schedule them alternating, A, B, A, B, .... and this solution also solves that as well.
-	for len(unschedulablePods) > 0 && len(unschedulablePods) != previousUnschedulableCount {
-		previousUnschedulableCount = len(unschedulablePods)
-		var newUnschedulablePods []*v1.Pod
-		for _, p := range unschedulablePods {
+	previouslyUnschedulable := 0
+	errors := map[*v1.Pod]error{}
+	var nodes []*Node
+	for len(pods) > 0 && len(pods) != previouslyUnschedulable {
+		previouslyUnschedulable = len(pods)
+		var unschedulable []*v1.Pod
+		for _, pod := range pods {
```

Comment:
I've been trying to avoid 'pod' as a name because of the pod package.  I don't feel strongly about it, but avoid it for that reason.

Reply:
Agreed. It's even more unnatural when we don't even import that package. Any developer who happens to have not come across one of those packages (or if its deprecated in the future), is unable to understand why we don't use the standard name.

Also -- We gotta rename that package.

Reply:
Yeah, +1 on renaming that to something else.